Add support for Ruby

sreeram-venkitesh opened the pull request:
@rohitpaulk I'm opening this as a draft PR so that you can review the progress as well.

I've added the starter template files for now. Will continue to work adding the rest of the code and diff files as I work my way through the Build your own Docker course with Go, and understand what we are doing in each step.

rohitpaulk commented:
@sreeram-venkitesh awesome!

I've enabled GitHub actions on this PR so you should see errors as they appear. Let me know once you have the starter templates ready and compiled (i.e. `compile_and_test.sh courses/docker ruby` passes), and I can deploy this on production so you can test the rest of the stages there. We have the ability to make languages as "alpha", so they won't be accessible to other users until we decide to.  

There's one [`go_lint` error](https://github.com/codecrafters-io/build-your-own-docker/actions/runs/3210949241/jobs/5249240845) that is unrelated to your changes, that I'm fixing in [this PR](https://github.com/codecrafters-io/build-your-own-docker/pull/11).

sreeram-venkitesh commented:
@rohitpaulk I think the `verify_templates_compiled` test failed since I hadn't created the files under `compiled_starters` directory. How can I create these files? Can I just copy the changes I made in `starter_templates`? I thought this would be generated somehow.

On a second viewing, I notice that the `README.md` file is not present in the files inside `starter_templates` and is only there inside `compiled_starters`. So I'm assuming I can copy the code files and add the appropriate content in the `README.md` file for Ruby, right?

rohitpaulk commented:
@sreeram-venkitesh the README here mentions how to develop courses: https://github.com/codecrafters-io/course-sdk#developing-courses. 

The "compile & test" command is what you're looking for 🙂 

> On a second viewing, I notice that the README.md file is not present in the files inside starter_templates and is only there inside compiled_starters.

It is present, just that it is common for all languages: https://github.com/codecrafters-io/build-your-own-docker/blob/main/starter_templates/README.md. When you run the compile_and_test command, it'll be automatically generated and placed inside `compiled-starters` 👍

sreeram-venkitesh commented:
@rohitpaulk Thanks! I've moved my version of the `Build your own Docker` repo to inside `courses/` in the `courses-sdk` repo locally so that I can run the tests and generate the files into my repo. I tried running compile and test, but I was getting an error with docker in the `Executing starter repo script` step:
```
docker: invalid reference format: repository name must be lowercase.
```
I'll look into this and let you know here once I've fixed it. Right now I'm going through the courses-sdk code and understanding how everything works.

rohitpaulk commented:
@sreeram-venkitesh interesting, haven't seen that one before. Might be worth confirming your docker version is the latest + ensuring no stray files exist anywhere that contain uppercase letters. (I'd do `git clean -fd`, which'll remove all untracked files). 

The same error isn't present on CI, here's what we're seeing there:
 
<img width="976" alt="Screenshot 2022-10-09 at 19 50 15" src="https://user-images.githubusercontent.com/3893573/194774429-a6a7726a-6a54-4be7-9c3f-c504042c5b56.png">

(Tried this on my local machine, and I see the same error as what CI reports)

rohitpaulk commented:
@sreeram-venkitesh just realized the GitHub actions requires me to approve workflows every time you push - I've added you as a collaborator so that this doesn't happen

rohitpaulk commented:
> I'll look into this and let you know here once I've fixed it. Right now I'm going through the courses-sdk code and understanding how everything works.

Awesome - curious to get to the bottom of this. If there's a way we catch the error early and display a friendly message to the user that'd be great

sreeram-venkitesh commented:
@rohitpaulk I was able to get around the `docker: invalid reference format: repository name must be lowercase` error locally by adding quotes around the options passed to `docker run` in the `assert_script_output` method in [starter_repo_tester.rb](https://github.com/codecrafters-io/course-sdk/blob/main/lib/starter_repo_tester.rb#L101), like so

```ruby
command = [
  "docker run",
  "-v '#{File.expand_path(tmp_dir, ENV["HOST_COURSE_SDK_PATH"])}:/app'",
  "-v '#{File.expand_path(tester_dir, ENV["HOST_COURSE_SDK_PATH"])}:/tester:ro'",
```

I had tried enclosing everything after `-v` in single quotes, after reading [this post](https://stackoverflow.com/questions/48522615/docker-error-invalid-reference-format-repository-name-must-be-lowercase#:~:text=Where%20if%20you%27re%20in%20the%20directory%20/home/user/Some%20Project%20Dir%2C%20that%20would%20define%20an%20anonymous%20volume%20/home/user/Some%20in%20your%20container%2C%20and%20try%20to%20run%20Project%3Alatest%20with%20the%20command%20Dir%3A/data%20image_ref.%20And%20the%20fix%20is%20to%20quote%20the%20argument%3A). I'm not sure if this was failing earlier due to something specific to my local dev setup, but worth looking into anyway.

Right now I'm getting the same error in `test_course_definition / test (ruby)` as in the GitHub Actions pipeline locally, which I'm looking into fixing now.

rohitpaulk commented:
@sreeram-venkitesh great find! Fixed in https://github.com/codecrafters-io/course-sdk/pull/9. 

Looks like you've got the GitHub actions tests passing now?

sreeram-venkitesh commented:
@rohitpaulk Yes, all the tests are passing now. 

I noticed that for all the languages other than Go, Build your own Docker only has the first stage right? I saw that we are showing Go's solutions for all the other languages as well. If this code is merged right now, I'm guessing the same would be the case for Ruby right? (for the people who are able to see languages marked as alpha that is)

rohitpaulk commented:
@sreeram-venkitesh yep! We'd show Go for the official solution.

As users try this out using Ruby (after it's in Beta), their solutions will automatically appear in the "Solutions" tab btw (released last week): 

<img width="1058" alt="Screenshot 2022-11-28 at 18 45 01" src="https://user-images.githubusercontent.com/3893573/204355998-d881353b-c44b-41d6-8944-24d792877e63.png">

rohitpaulk commented:
@sreeram-venkitesh so this is good to review? (PR is still a draft - let me know when you want me to take a look 🙂)

sreeram-venkitesh commented:
@rohitpaulk I'm done from my end, please take a look and let me know if I can proceed in the same manner for the rest of the parts.

rohitpaulk commented:
Hey @sreeram-venkitesh! Sorry, this got lost in a sea of notifications - taking a look soon 🙂 

> Is there any other way I can test if the solution is correct other than by running the tests locally?

The same local tests are run on CI too (and it looks like those are passing), so I think we're good (will review in detail soon)!

rohitpaulk commented:
Quick update - going to take a look at this tomorrow (sorry, past week was super busy)

rohitpaulk commented on solutions/ruby/01-init/definition.yml:
```diff
@@ -0,0 +1,5 @@
+author_details:
+  name: Sreeram Venkitesh 
+  profile_url: https://github.com/sreeram-venkitesh
+  avatar_url: https://github.com/sreeram-venktiesh.png
```
```suggestion
  avatar_url: https://github.com/sreeram-venkitesh.png
```

rohitpaulk commented on starter_templates/ruby/app/main.rb:
```diff
@@ -0,0 +1,16 @@
+# You can use print statements as follows for debugging, they'll be visible when running tests.
```
Let's switch the order of statements a bit here. Let's place `require "open3"` at the top of the file (can be commented / uncommented, either works). That's what we've been doing with other languages, example: 

<img width="831" alt="Screenshot 2022-12-13 at 15 36 02" src="https://user-images.githubusercontent.com/3893573/207376580-ab9090e7-d5b1-47ec-b4fd-389251d01f59.png">

For context: the most common usage here is that a user will just blindly uncomment the block and then run tests. They'll _then_ see "Logs from your program will appear here!" in the test output and might remove that line in later stages.

rohitpaulk commented on starter_templates/ruby/app/main.rb:
```diff
@@ -0,0 +1,16 @@
+# require "open3"
+#
+# command = ARGV[2]
+# args = ARGV[3..ARGV.length]
```
Is the `..ARGV.length` needed? Can you just do `ARGV[3..]`?

rohitpaulk commented on starter-repository-definitions.yml:
```diff
@@ -103,3 +103,19 @@
+    - source: starter_templates/ruby/your_docker.sh
+      target: your_docker.sh
+  template_attributes:
+    required_executable: "ruby (3.0)"
```
Let's make this 2.7 for now (we don't support 3.x yet, and that can be a separate PR)

rohitpaulk commented:
Deploying this in a couple of mins...

rohitpaulk commented:
@sreeram-venkitesh can you head over to https://app.codecrafters.io/courses/docker?fresh=true and see if Ruby is now listed for your account?

rohitpaulk commented:
@sreeram-venkitesh wouldn't worry about further "official" solutions, we don't need those to release this to the public 🙂  Just get through the challenge using your account, and your solutions will automatically appear in the "solutions" tab: 

<img width="1054" alt="Screenshot 2022-12-13 at 16 25 58" src="https://user-images.githubusercontent.com/3893573/207388584-7bfca0b1-0d91-4508-8295-37334f9ce4fe.png">

If you get through till stage 5 (process isolation), I think it's safe to release for others to try. Stage 6 only requires wiring up with the API, and that's straightforward in Ruby. The part I'm unsure about is process isolation (using PID namespaces).

sreeram-venkitesh commented:
Ah okay, so once I attempt the course in Ruby, my code would show up in the `Solutions` tab and even though official solutions aren't there for Ruby, that's enough for releasing it to the public, right? Anyway I'll let you know here once I'm done with all the stages.

rohitpaulk commented:
> Ah okay, so once I attempt the course in Ruby, my code would show up in the Solutions tab and even though official solutions aren't there for Ruby, that's enough for releasing it to the public, right? Anyway I'll let you know here once I'm done with all the stages.

Yep, exactly!
